Run wrench on travis CI with OSMesa

glennw opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/649)
<!-- Reviewable:end -->

kvark commented on wrench/Cargo.toml:
```diff
@@ -33,3 +33,7 @@ dwrote = "0.1.2"
 [target.'cfg(target_os = "linux")'.dependencies]
 font-loader = "0.1.0"
+
+[target.'cfg(any(target_os = "linux", target_os = "macos"))'.dependencies]
```
given that we use it for automated tests only, can we hide this dependency behind a feature gate?

glennw replied:
Moved this to the 'headless' feature.

kvark commented on wrench/run_tests.py:
```diff
@@ -0,0 +1,63 @@
+        os.environ["GALLIUM_DRIVER"] = "softpipe"
+
+
+set_osmesa_env('../target/release/')
```
is our executable built by `cargo test` guaranteed to be in `release` folder?

glennw replied:
I explicitly build the release version in the travis script - perhaps there's a better way to do this though.

kvark commented on wrench/src/main.rs:
```diff
@@ -75,31 +86,161 @@ impl ThingKind {
+    fn get_proc_address(s: &str) -> *const c_void {
+        let c_str = CString::new(s).expect("Unable to create CString");
+        unsafe {
+            mem::transmute(osmesa_sys::OSMesaGetProcAddress(c_str.as_ptr()))
```
can we use a simple `X as *const _` here instead of `transmute`?

glennw replied:
The OSMesa bindings return an Option<fn> here so I think a transmute is required here due to the interface mismatch. We could perhaps make these match in the future and remove the transmute?

kvark replied:
Sounds good! No need to have an extra `unsafe` where it's not required

kvark commented on wrench/src/main.rs:
```diff
@@ -175,97 +320,138 @@ fn main() {
-                if show_help {
-                    wrench.show_onscreen_help();
-                }
+            let thing = thing.thing();
```
![thing](http://schmoesknow.com/wp-content/uploads/2015/10/TheThing.jpg)

kvark commented on wrench/src/main.rs:
```diff
@@ -175,97 +320,138 @@ fn main() {
-                            _ => ()
+                glutin::Event::KeyboardInput(kstate, _scan_code, maybe_vk) => {
+                    if kstate == ElementState::Pressed {
+                        if let Some(vk) = maybe_vk {
```
nit: we could combine both of these `if`s

glennw replied:
Unrelated to this PR, so leaving for now.

kvark commented on wrench/src/main.rs:
```diff
@@ -175,97 +320,138 @@ fn main() {
+        let mut pixels = gl::read_pixels(0, 0,
+                                         size.width as gl::GLsizei,
+                                         size.height as gl::GLsizei,
+                                         gl::RGB, gl::UNSIGNED_BYTE);
```
would that incur a software format conversion from RGBA to RGB?

glennw replied:
Perhaps - I'm not entirely sure. But it's only done once when writing the screenshot in test mode, so probably not a big deal.

glennw commented:
This is a bit untidy, but it now runs wrench in headless / osmesa mode on travis (mac + linux), writes a screenshot, and the md5sum of that png matches on both CI machines, and also on my local machine when I check the image output visually.

@jrmuizel Hopefully this is enough for you to continue your work on adding basic test support. Let me know if anything isn't working or needs some additional functionality.

@kvark I've addressed most of the comments you left, but if you have time to take another look over it and see what else needs cleaning up, we could perhaps get this part merged soon-ish.

glennw commented on .travis.yml:
```diff
@@ -8,8 +8,15 @@ notifications:
     packages:
       - libgles2-mesa-dev
+      - llvm-3.8
+      - llvm-3.8-dev
```
Fixed

glennw commented on wrench/src/main.rs:
```diff
@@ -21,18 +21,29 @@ extern crate serde;
 extern crate serde_json;
 extern crate crossbeam;
 
+#[cfg(any(target_os = "linux", target_os = "macos"))]
```
Added the headless feature.

nox commented:
@glennw Seems that Travis is happy now.

kvark commented on wrench/src/main.rs:
```diff
@@ -21,18 +21,29 @@ extern crate serde;
+use std::ptr;
 use webrender_traits::*;
 
+#[cfg(any(target_os = "linux", target_os = "macos"))]
```
should use `headless` feature

kvark commented on wrench/src/main.rs:
```diff
@@ -75,31 +86,161 @@ impl ThingKind {
+        match *self {
+            WindowWrapper::Window(ref window) => {
+                unsafe {
+                    window.make_current().ok();
```
I don't think it makes sense to ignore the returned `Err` here, we better just do `unwrap` or `expect` instead.

kvark commented on wrench/src/main.rs:
```diff
@@ -175,97 +320,138 @@ fn main() {
+                events.push(glutin::Event::Awakened);
+            }
+            WindowWrapper::Window(ref window) => {
+                for event in window.poll_events() {
```
nit: could do shorter:
```rust
let events = match window {
  WindowWrapper::Headless(..) => {
    vec![glutin::Event::Awakened]
  }
  WindowWrapper::Window(ref window) => {
    window.poll_events().collect()
  }
};
```

glennw commented:
@kvark Thanks for the review, addressed those comments and squashed.

kvark commented:
Thanks!
I suppose this error at the end is expected?
> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libcore/result.rs:837
note: Run with `RUST_BACKTRACE=1` for a backtrace.
md5 = 31ca82c0aadec2c0ebfd0e3f95510c9c
/Users/travis/build.sh: line 57: shell_session_update: command not found

Either way, we can follow up with more work, at least Travis is green.
@bors-servo r+

bors-servo commented:
:pushpin: Commit 55af167 has been approved by `kvark`

<!-- @bors-servo r=kvark 55af167c3477439094997814e4937f32b5775130 -->

bors-servo commented:
:hourglass: Testing commit 55af167 with merge 840d707...

bors-servo commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/servo/webrender/builds/187299268)

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/187299268)

bors-servo commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

kvark commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 55af167 with merge b956a28...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/187332286)
